### DISCUSSION TITLE
Format for faces for PolyData

### DISCUSSION THREAD

**Discussion by MatthewFlamm** [pyvista/pyvista#2799] [Ideas]
I find this format to be non-intuitive and hard to construct:

```py
points = [[0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 0], [0, 0, 1], [1, 0, 1], [1, 1, 1], [0, 1, 1.0]]
faces = [4, 0, 1, 2, 3, 4, 0, 1, 5, 4, 4, 1, 2, 6, 5, 4, 2, 3, 7, 6, 4, 3, 1, 4, 7, 4, 4, 5, 6, 7]
mesh = pyvista.PolyData(points, faces)
```

I would find the following format preferable:

```py
points = [[0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 0], [0, 0, 1], [1, 0, 1], [1, 1, 1], [0, 1, 1.0]]
faces = [[0, 1, 2, 3], [0, 1, 5, 4], [1, 2, 6, 5], [2, 3, 7, 6], [3, 1, 4, 7], [4, 5, 6, 7]]
```

There is no loss of information in the second format, but it is much more human readable.  It is also easier to construct in code.

**Comment by tkoyama010** [comment]
I agree. This is where PyVista beginners always make mistakes. I am +1 to fix it, but it's difficult to ensure backward compatibility if we make it the default.

**Reply by MatthewFlamm** [reply]
Same thing for `UnstructuredGrid`.  Currently these classes take in a flat 1D list and pass it to vtk.  We can check for 2D lists (or ndarrays) and then convert it to the form vtk wants.  If my suggestion requires copying the data to make it in the vtk form, then we should never remove the current functionality.

Edit: What I'm suggesting is have two options: one that is more convenient (my suggestion) and one that is memory/cpu efficient (current state?).

**Reply by darikg** [reply]
In addition to making the PolyData/etc constructors more convenient, we'd also want to change the faces/lines/etc properties to return reshaped arrays, which seems even harder to do backwards-compatibly.

**Reply by whophil** [reply]
On properties - how should the `faces` property return a reshaped array when the mesh has faces of different sizes?

**Reply by darikg** [reply]
Good point -- I guess instead we could have convenience methods .faces_array(), etc., and the properties could remain as is. I retract my comment!
